### ISSUE TITLE
Generalisation to different greps/integration question

### ISSUE THREAD

**Issue by minad** [gexplorer/ag-popup#1]
Hi Eder, 

I saw [your post on reddit](https://www.reddit.com/r/emacs/comments/kvvgxu/counselagpopup_the_power_of_searching_with_ag/) and I appreciate the efforts to generalize your original package, such that it could used together with different grep backends.

I wonder if it would be possible to generalize this package in order to support different greps? I am in particular interested in grep, git-grep and ripgrep.

It would be great if I could glue your package together with the `consult-grep` functions provided by my [Consult](https://github.com/minad/consult) package. In Consult I am splitting the minibuffer input entered by the user as follows `#grep-regexp -- grep-options#filter-string`. You can edit all parts of this input expressions and the search results will be updated on the fly. Is it possible to setup ag-popup somehow that it stays alive and updates the `grep-options` part of the minibuffer input?

Daniel

**Comment by gexplorer** [comment]
Hi Daniel,

I like the idea of integrating with consult. I started with ag because I've been and **ivy** user for a long time and I'm used to **counsel-ag**. Anyway I'm preparing a [grep-popup](https://github.com/gexplorer/grep-popup) since **ag** arguments are not the same as **grep** arguments.
 
If you're interested I'd love to see if there's a way to make a transient popup work with **consult**.

**Comment by minad** [comment]
> I like the idea of integrating with consult. I started with ag because I've been and ivy user for a long time and I'm used to counsel-ag.

Great to hear! Ag makes sense in your case. But for me it has been pretty much superseded by Ripgrep. For this reason I also didn't add Ag support to Consult. I still have Grep and Git-grep support because of the ubiquity of those commands and it does not cost much to add commands.

> Anyway I'm preparing a grep-popup since ag arguments are not the same as grep arguments.

Is it possible to have some kind of code sharing between grep-popup/ripgrep-popup/ag-popup or is there no other way than replicating everything since the options differ too much? I am a fan of unified solutions :)

> If you're interested I'd love to see if there's a way to make a transient popup work with consult.

I must admit I didn't try your package yet, but from what I saw it seems that you open this popup, select the options and then pass them to counsel-ag? After that you cannot change the options anymore? Ideally I would like to have something which allows me to change options on the go, since the options are part of the filter expression, as I mentioned before. Maybe transient is the wrong tool to achieve that since it takes over the transient keymap? Maybe in my case a small help popup without any capabilities to modify the options would also do it. But I can also imaging putting your transient popup at some minibuffer keybinding, such that I can summon it can be summoned at any time. Then the popup simply manipulates the minibuffer content, which will then automatically trigger a search with the updated settings.

**Comment by gexplorer** [comment]
I agree that if you want a popup for this the way to go would be toggling it with a keybinding in the minibuffer. The creation of the actual popups would be the easy part, I can create them easily enough. The trickier part would be to find a way to make them work with consult, but from what I've been able to see from the code, it seems well organized.

Just for fun I created a [dirty hack](https://github.com/minad/consult/pull/170) to give it a try.
